Handle duplicated dimensions

jamesalster opened the pull request:
This is a PR addressing issue #18 . I believe the problem can be addressed without changing any of the maths. However, it turned out that:

* the uniqueness of dimensions in `DimIndices` objects was implicitly assumed by a lot of downstream methods, which needed to be rewritten
* it's also necessary to perform more checks on `ArrayMargins` objects, since there is a lot of potential for issues when dimensions are repeated inconsistently

Both of those made this rather a large PR, and I've broken it up into several commits to try and make things clearer. 

While it implements the required functionality and passes the tests, I'm sure the code could be improved, and I'm more than happy to work on this further - but I thought it had reached a point where any feedback would be very welcome.

vankesteren commented:
Ah great work! I will try to review it as soon as I have some time; my review will require running the code I think.

vankesteren commented on src/DimIndices.jl:
```diff
@@ -22,9 +22,12 @@ struct DimIndices
-        # uniqueness
-        if !allunique(vcat(idx...)) 
-            error("Some dimensions were duplicated.")
+        # uniqueness between sets of indices 
+        if !allunique(sort.(idx))
+            error("Some sets of dimensions were duplicated, e.g. [[1,2], [2,1]].")
+        # uniqueness within sets of indices
+        elseif !all(allunique, idx)
+            error("Some dimensions were duplicated within a set, e.g. [[2,1,2], [3]].")
```
Nice solution

vankesteren commented on src/ipf.jl:
```diff
@@ -51,23 +51,29 @@ function ipf(X::AbstractArray{<:Real}, mar::ArrayMargins; maxiter::Int = 1000, t
     end
     array_size = size(X)
-    if array_size != size(mar)
+    if array_size != size(mar) #calling size(mar) implicitly checks dimension sizes between arrays
```
This implicit check will change, see comments above

vankesteren commented on src/ipf.jl:
```diff
@@ -51,23 +51,29 @@ function ipf(X::AbstractArray{<:Real}, mar::ArrayMargins; maxiter::Int = 1000, t
         X /= sum(X)
         mar = proportion_transform(mar)
+
+        #recheck proportions across margins that appear more than once
+        if !check_margin_totals(mar; tol = tol)
+            throw(DimensionMismatch("Margin proportions inconsistent across dimensions"))
+        end
```
Is this necessary? Is this not guaranteed? I understand why you'd want to double check, but maybe this is superfluous

vankesteren commented on src/ipf.jl:
Wow, there is way less change in the ipf function than I thought there would be! nice!

vankesteren commented on test/runtests.jl:
```diff
@@ -43,6 +53,43 @@ end
+    target_23 = fill(2, (3,4))
+    di = DimIndices([[1,3], [2,1]]) #incorrect should be [[1,3], [2,3]]
+    m = ArrayMargins([target_13, target_23], di)
+    @test_throws DimensionMismatch size(m)
```
This will change if we put the check in the constructor; then the @test_throws DimensionMismatch should be one line higher 👍

vankesteren commented on test/runtests.jl:
```diff
@@ -80,6 +127,16 @@ end
+    f5 = reshape(10:13, 2, 2)
+    # test error
+    fac8 = ArrayFactors([f2, f4, f2'], di)
+    @test_throws DimensionMismatch size(fac8)
```
Same here

vankesteren commented on src/ArrayMargins.jl:
```diff
@@ -111,6 +119,37 @@ function isconsistent(AM::ArrayMargins; tol::Float64 = eps(Float64))
+    )
+
+    #loop over these subsets, and check marginal totals are equal in every array margin where they appear
+    check = true
+    for dd in shared_subsets
+        margin_totals = Vector{Array{Float64}}()
+        for i in 1:length(AM.am)
+            if issubset(dd, AM.di.idx[i])
+                dims_to_sum = findall(x -> !in(x, dd), AM.di.idx[i])
+                margin_total = dropdims(sum(AM.am[i]; dims = dims_to_sum); dims = Tuple(dims_to_sum))
+                #reshape
+                remaining_dims = filter(d -> d in dd, AM.di.idx[i])
+                perm_order = [findfirst(==(d), remaining_dims) for d in dd]
+                push!(margin_totals, permutedims(margin_total, perm_order))
+            end
+        end
+        if !all(x -> isapprox(x, margin_totals[1]; atol = tol), margin_totals)
+            @warn "Margin totals do not match across array margin(s): $dd"
+            check = false
+        end
```
Yeah, this looks so complicated but I guess it's necessary...

vankesteren reviewed (changes requested):
Awesome work! Actually, this is not as much difference as I thought it would be. I had expected much more change in the actual algorithm. 

The main change I'm requesting is to put the new checks for the ArrayMargins and ArrayFactors objects directly in their constructors, rather than in the `size()` method.

For the rest there are only minor comments/suggestions

jamesalster commented:
Many thanks for the detailed review! I've committed to make the minor changes, and I'll make a bigger commit later today or possibly tomorrow for the `size(::ArrayMargins)` change, moving the check to an inner constructor.

I agree that some of the methods are nasty, and I thought of a neater way to do the margins sum check that I *think* will work, I'll do this when I make the commit later.

vankesteren commented:
Oh yeah, and don't forget to put yourself in the authors section in the Project.toml!

jamesalster commented:
Ok, I think we're good for another review. I have:
* Moved the size() check on ArrayMargins and ArrayFactors into an inner constructor
* Changed that check to be a bit clearer (I did think of an even nicer way of doing this with `get!()` but that would require depending on OrderedCollections.jl
* Renamed the array margins check, and changed the logic to move the reshaping into a new function `align_arrays()` (maybe this should actually be `align_margins()`

This last I think makes it clearer, even if there are more lines. I also thought that we might be able to use this function or a modified version of it to pre-sort the array factors in `ipf()` before we start the main loop, and take the equivalent logic out of the inner loop there. But that is definitely for a separate PR, I think.

vankesteren reviewed (approved):
I have a very minor comment, but in general I'd say this PR is basically finished!

vankesteren commented:
My plan for now: 

- merge this PR once you think it's ready
- make a new PR with version bump to 0.4.0, which also includes a slight update to the docs to show the new feature
- register the new version
- make 0.5.0 focus on performance increase (the adjust!() function and pre-computed margin alignment

vankesteren commented:
Alright, yes, merging. Sounds good if you want to pick this up! Then I'll focus on the other performance thing. I will make issues and assign you 👍
